// eslint-plugin-gridsome/dist/lib/utils/get-code-wrap-indent-info.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCodeWrapIndentInfo = void 0;
const indent = {
    space: " ",
    tab: "\t",
    none: "",
};
exports.getCodeWrapIndentInfo = (mergedPrettierOptions, eslintOption) => {
    var _a, _b;
    const defaultInfo = {
        indentRepeatTime: 0,
        indentChar: indent.space,
    };
    const prettierOptions = mergedPrettierOptions;
    const tabWidth = (eslintOption === null || eslintOption === void 0 ? void 0 : eslintOption.tabWidth) || prettierOptions.tabWidth;
    const useTabs = (_a = eslintOption === null || eslintOption === void 0 ? void 0 : eslintOption.useTabs) !== null && _a !== void 0 ? _a : prettierOptions.useTabs;
    const vueIndentScriptAndStyle = (_b = eslintOption === null || eslintOption === void 0 ? void 0 : eslintOption.vueIndentScriptAndStyle) !== null && _b !== void 0 ? _b : prettierOptions.vueIndentScriptAndStyle;
    const result = {
        indentRepeatTime: tabWidth,
        indentChar: defaultInfo.indentChar,
    };
    if (useTabs) {
        result.indentRepeatTime = 1;
        result.indentChar = indent.tab;
    }
    if (!vueIndentScriptAndStyle) {
        result.indentRepeatTime = 0;
        result.indentChar = indent.none;
    }
    return Object.assign(Object.assign({}, defaultInfo), result);
};
