// eslint-plugin-gridsome/dist/lib/rules/require-g-link-to.js
"use strict";
const utils_1 = require("../utils");
const defaultOptions = [{}];
const statementTo = (node) => !utils_1.hasAttribute(node, "to") && !utils_1.hasDirective(node, "bind", "to");
const statementHref = (node) => !utils_1.hasAttribute(node, "href") && !utils_1.hasDirective(node, "bind", "href");
module.exports = utils_1.createRule({
    name: "require-g-link-to",
    meta: {
        docs: {
            description: "Require v-bind:to or to of `<g-link>` elements",
            category: "Possible Errors",
            recommended: false,
        },
        type: "problem",
        messages: {
            requireGLinkTo: "Expected '<g-link>' elements to have 'v-bind:to', 'to' or 'v-bind:href', 'href'.",
        },
        schema: [],
    },
    defaultOptions,
    create(context) {
        return utils_1.defineTemplateBodyVisitor(context, {
            "VElement[name='g-link']"(node) {
                if (statementTo(node) && statementHref(node)) {
                    context.report({
                        node,
                        loc: node.loc,
                        messageId: "requireGLinkTo",
                    });
                }
            },
        });
    },
});
