// eslint-plugin-gridsome/dist/lib/rules/require-g-image-src.js
"use strict";
const utils_1 = require("../utils");
const defaultOptions = [{}];
module.exports = utils_1.createRule({
    name: "require-g-image-src",
    meta: {
        docs: {
            description: "Require v-bind:src or src of `<g-image>` elements",
            category: "Possible Errors",
            recommended: false,
        },
        type: "problem",
        messages: {
            requireGImageSrc: "Expected '<g-image>' elements to have 'v-bind:src' or 'src'.",
        },
        schema: [],
    },
    defaultOptions,
    create(context) {
        return utils_1.defineTemplateBodyVisitor(context, {
            "VElement[name='g-image']"(node) {
                if (!utils_1.hasAttribute(node, "src") && !utils_1.hasDirective(node, "bind", "src")) {
                    context.report({
                        node,
                        loc: node.loc,
                        messageId: "requireGImageSrc",
                    });
                }
            },
        });
    },
});
